src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -299,28 +309,28 @@ private BoundExpression VisitCollectionBuilderCollectionExpression(BoundCollecti
             return result;
         }
 
-        internal static bool ShouldUseRuntimeHelpersCreateSpan(BoundCollectionExpression node, TypeSymbol elementType)
+        internal static bool IsAllocatingRefStructCollectionExpression(BoundCollectionExpressionBase node, TypeSymbol targetType, TypeSymbol? elementType, CSharpCompilation compilation)
         {
-            Debug.Assert(node.CollectionTypeKind is
-                CollectionExpressionTypeKind.ReadOnlySpan or
-                CollectionExpressionTypeKind.CollectionBuilder);
+            return targetType.IsRefLikeType
+                && !node.Elements.IsEmpty
+                && elementType is not null
+                && !(targetType.OriginalDefinition.Equals(compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T)) && ShouldUseRuntimeHelpersCreateSpan(node, elementType))
```

Comment:
>targetType.OriginalDefinition.Equals(compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T)) [](http://example.com/codeflow?start=21&length=103)

Will this be false for a `[CollectionBuilder]` type, even in the case where lowering will use the `ShouldUseRuntimeHelpersCreateSpan()` optimization? For instance, for a case such as `MyCollection<int> c = [1, 2, 3];`?

Reply:
It could be. Based on this and other concerns about the actionability of the diagnostic I am going to go ahead with limiting it to Span/ROS for now, and not issue a diagnostic for conversion to user-defined ref struct.